polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp

```diff
@@ -107,6 +109,179 @@ struct RaiseKernelName : public OpRewritePattern<LLVM::AddressOfOp> {
+    rewriter.create<LLVM::BrOp>(invoke->getLoc(),
+                                invoke.getNormalDestOperands(),
+                                invoke.getNormalDest());
+  }
+};
+
+struct BufferTypeTag {
+
+  static llvm::StringRef getTypeName() { return "class.sycl::_V1::buffer"; }
+
+  static mlir::Type getTypeFromConstructor(CallOpInterface constructor) {
+    CallInterfaceCallable callableOp = constructor.getCallableForCallee();
+    StringRef constructorName =
+        callableOp.get<SymbolRefAttr>().getLeafReference();
+
+    auto demangledName = llvm::demangle(constructorName);
+
+    // Try to determine the dimensions of the buffer by parsing the template
+    // parameter from the demangled name of the constructor.
+    llvm::Regex bufferTemplate("buffer<.*, ([0-9]+)");
+    llvm::SmallVector<StringRef> matches;
+    bool regexMatch = bufferTemplate.match(demangledName, &matches);
+    unsigned dimensions = 1;
+    if (regexMatch)
+      dimensions = std::stoul(matches[1].str());
+
+    // FIXME: There's currently no good way to obtain the element type of the
+    // buffer from the constructor call (or allocation). Parsing it from the
+    // demangled name, as done for 'dimensions' above, would require translation
+    // from C++ types to MLIR types, which is not available here.
+    Type elemTy = LLVM::LLVMVoidType::get(constructor->getContext());
+
+    return sycl::BufferType::get(constructor->getContext(), elemTy, dimensions);
+  }
+};
+
+struct BufferInvokeConstructorPattern
+    : public RaiseInvokeConstructorBasePattern<BufferTypeTag> {
+  using RaiseInvokeConstructorBasePattern<
+      BufferTypeTag>::RaiseInvokeConstructorBasePattern;
+};
+
+struct AccessorTypeTag {
+
+  static llvm::StringRef getTypeName() { return "class.sycl::_V1::accessor"; }
+
+  static mlir::Type getTypeFromConstructor(CallOpInterface constructor) {
+    CallInterfaceCallable callableOp = constructor.getCallableForCallee();
+    StringRef constructorName =
+        callableOp.get<SymbolRefAttr>().getLeafReference();
+
+    auto demangledName = llvm::demangle(constructorName);
+
+    // Try to determine the parameters of the accessor by parsing the template
+    // parameter from the demangled name of the constructor.
+    llvm::Regex accessorTemplate(
+        "accessor<.*, ([0-9]+), \\(sycl::_V1::access::mode\\)([0-9]+), "
+        "\\(sycl::_V1::access::target\\)([0-9]+)");
+    llvm::SmallVector<StringRef> matches;
+    bool regexMatch = accessorTemplate.match(demangledName, &matches);
+    unsigned dimensions = 1;
+    unsigned target = 2014;
+    unsigned mode = 1026;
+    if (regexMatch) {
+      dimensions = std::stoul(matches[1].str());
+      mode = std::stoul(matches[2].str());
+      target = std::stoul(matches[3].str());
+    }
+
+    sycl::AccessMode accessMode = static_cast<sycl::AccessMode>(mode);
+    sycl::Target accessTarget = static_cast<sycl::Target>(target);
```

Comment:
We have `mlir::sycl::symbolizeAccessMode` and `mlir::sycl::symbolizeTarget` equivalent to these and with bounds checking.